lib/pbio/src/broadcast.c

```diff
@@ -0,0 +1,236 @@
+
+    // Clear transmitted signal.
+    transmit_signal.value.size = 0;
+
+    // Clear number of signals we are scanning for.
+    num_scan_signals = 0;
+}
+
+void pbio_broadcast_start(void) {
+
+    if (!transmit_signal.process_running) {
+        process_start(&pbio_broadcast_process);
+        transmit_signal.process_running = true;
+    }
+}
+
+void pbio_broadcast_stop(void) {
+
+    if (transmit_signal.process_running) {
+        // stop broadcast process
+        process_exit(&pbio_broadcast_process);
+        transmit_signal.process_running = false;
+    }
+
+    // stop scanning
+    pbdrv_bluetooth_broadcast_start_scan(false);
+
+    // stop advertising
+    pbdrv_bluetooth_stop_data_advertising();
+}
+
+pbio_error_t pbio_broadcast_register_signal(uint32_t hash) {
+    // Check if there are any slots left.
+    if (num_scan_signals == PBIO_CONFIG_BROADCAST_NUM_SIGNALS) {
+        return PBIO_ERROR_INVALID_ARG;
+    }
+
+    // Go through all signals to see if this hash is already in use.
+    for (uint8_t i = 0; i < num_scan_signals; i++) {
+        if (received_signals[i].hash == hash) {
+            return PBIO_ERROR_INVALID_ARG;
+        }
+    }
+
+    // Get the next available signal and increment the number of used slots.
+    pbio_broadcast_received_t *signal = &received_signals[num_scan_signals++];
+
+    // Reset the signal.
+    signal->hash = hash;
+    signal->index = 0;
+    signal->size = 0;
+
+    return PBIO_SUCCESS;
+}
+
+void pbio_broadcast_receive(uint32_t hash, uint8_t **payload, uint8_t *size) {
+    for (uint8_t i = 0; i < num_scan_signals; i++) {
+
+        pbio_broadcast_received_t *s = &received_signals[i];
+
+        // Return if there is a match.
+        if (s->hash == hash) {
+            *payload = s->payload;
+            *size = s->size;
+            return;
+        }
+    }
+
+    // No signal registered for this hash, so size is 0.
+    *size = 0;
+}
+
+void pbio_broadcast_transmit(uint32_t hash, const uint8_t *payload, uint8_t size) {
+
+    // Cut off payloads that are too long.
+    if (size > PBIO_BROADCAST_MAX_PAYLOAD_SIZE) {
+        size = PBIO_BROADCAST_MAX_PAYLOAD_SIZE;
+    }
+
+    uint8_t old_payload_size = transmit_signal.value.size - PBIO_BROADCAST_META_SIZE;
+    uint32_t time_now = pbdrv_clock_get_ms();
+
+    // If we've sent the same message recently, there is nothing we need to do.
+    if (transmit_signal.hash == hash &&
+        old_payload_size == size &&
+        time_now - transmit_signal.timestamp < PBIO_BROADCAST_DELAY_REPEAT_MS &&
+        !memcmp(transmit_signal.payload, payload, size)) {
+        return;
+    }
+
+    // If new the message is on the same signal, increment the index.
+    if (transmit_signal.hash == hash) {
+        transmit_signal.index++;
+    } else {
+        // Otherwise reset the index.
+        transmit_signal.index = 0;
+        transmit_signal.hash = hash;
+    }
+
+    // Copy the payload.
+    transmit_signal.length = size - 1;
```

Comment:
Why length does not include PBIO_BROADCAST_META_SIZE?
In my opinion, it should be added to the length. When I receive the advertising data I miss exactly 9 bytes.
```suggestion
    transmit_signal.length = size + PBIO_BROADCAST_META_SIZE - 1;
```

Reply:
@NStrijbosch: Feel free to comment on this. It may be a while before we can get back to this.

Reply:
@pbochynski You are correct! I just pushed an update with your fix. 

In the current implementation in pybricks this value is never checked, so from pybricks hub to pybricks hub communication this value didn't really matter, so I never noticed the mistake.